bokeh/charts/chart.py

```diff
@@ -39,39 +40,38 @@
 from ..resources import INLINE
 from ..util.notebook import publish_display_data
 from ..util.serialization import make_id
+from ..util.future import with_metaclass
+from ..model import Viewable
+from ..themes import Theme
 
 #-----------------------------------------------------------------------------
 # Classes and functions
 #-----------------------------------------------------------------------------
 
-def _make_method(prop_name):
-    def method(self, value):
-        setattr(self._options, prop_name, value)
-        return self
-    method.__doc__ = """ Chained method for %s option.
-    """ % prop_name
-    return method
+Scale = enumeration('linear', 'categorical', 'datetime')
 
-def _chained_options(opts_type):
-    def wrapper(cls):
-        orig_init = cls.__init__
+class ChartsTheme(Theme):
+    def apply_to_model(self, model):
+        """Apply this theme to a model. Don't call this method directly,
+        instead set the theme on the Document the model is a part of."""
 
-        cls_props = set(cls.properties())
+        if not isinstance(model, Chart):
+            raise ValueError("ChartsTheme should be only used on Chart objects \
+            but it's being used on %s instead." % model)
 
-        def __init__(self, *args, **kwargs):
-            self._options = opts_type(**kwargs)
-            orig_init(self)
+        attrs = {}
+        for k in model.properties_with_values(include_defaults=True):
+            if hasattr(self, k):
+                attrs[k] = getattr(self, k)
 
-        cls.__init__ = __init__
+        if attrs:
+            self._by_class_cache = {}
+            self._json = {'attrs' : { 'Chart' : attrs }}
 
-        for prop_name in opts_type.properties():
-            if prop_name not in cls_props:
-                setattr(cls, prop_name, _make_method(prop_name))
+        model.apply_theme(self._for_class(model.__class__))
```

Comment:
If you are doing this I think you could just use a dict - you don't need a subtype of Theme or a special class?
`defaults = {}` ought to work as far as I can see above, then `model.apply_theme(defaults)` below.

However, it won't really work right as-is. The issue is that the theme is set in two other places.
 * in `Model.__init__` we set a default theme which is empty right now but in theory could be configurable, maybe `BOKEH_THEME=dark.yaml` or something like that. 
 * when the Chart is added to a Document, it will switch to `Document.theme` overwriting the above

The solution I suppose could be to add another layer in HasProps; take this code for theme:
https://github.com/bokeh/bokeh/blob/master/bokeh/properties.py#L834-L880

and parameterize it so the `__themed_values__` name is passed in. So like:
```python
 def _apply_theme_layer(self, property_values, attr):
    # all the current apply_theme code using attr instead of `__themed_values__`
 def apply_theme(self, property_values):
    self._apply_theme_layer(property_values, '__themed_values__')
```

Then over here you'd have to combine all "theme layers": https://github.com/bokeh/bokeh/blob/master/bokeh/properties.py#L350-L352

So the layer you're creating here I guess could be called "dynamic_defaults" (defaults set by the app developer at runtime?) or something like that?
```
def apply_dynamic_defaults(self, property_values):
  self._apply_theme_layer(property_values, '__dynamic_defaults__')
```

I don't know. It seems like a lot of conceptual/documentation/code complexity relative to the benefits, to me. I'd be very tempted to drop this feature or try to replace it with one that isn't Chart-specific.